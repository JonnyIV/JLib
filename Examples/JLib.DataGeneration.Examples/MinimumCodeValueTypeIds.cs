// Third party packages
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Xunit.Abstractions;
using Snapshooter.Xunit;
using Xunit;

// required JLib packages
using JLib.DependencyInjection;
using JLib.Exceptions;
using JLib.Helper;
using JLib.Reflection;
using JLib.AutoMapper;

// referenced example setup code
using JLib.DataGeneration.Examples.Setup.Models;
using JLib.DataGeneration.Examples.Setup.SystemUnderTest;

namespace JLib.DataGeneration.Examples;
public sealed class MinimumCodeValueTypeIds : IDisposable
{
    /*************************************************************\
    |                       Data Packages                         |
    \*************************************************************/
    public class CustomerDp : DataPackage
    {
        public CustomerId CustomerId { get; set; } = null!;
        public CustomerDp(ShoppingServiceMock shoppingService, IDataPackageManager packageManager) : base(packageManager)
        {
            shoppingService.AddCustomer(new(GetInfoText(nameof(CustomerId)))
            {
                Id = CustomerId
            });
        }
    }

    /*************************************************************\
    |                           Setup                             |
    \*************************************************************/
    private readonly List<IDisposable> _disposables = new();
    private readonly ShoppingServiceMock _shoppingService;

    public MinimumCodeValueTypeIds(ITestOutputHelper testOutputHelper)
    {
        var exceptions = ExceptionBuilder.Create("setup");
        var loggerFactory = new LoggerFactory().AddXunit(testOutputHelper);
        
        var serviceCollection = new ServiceCollection()
            .AddTypeCache(out var typeCache, exceptions, loggerFactory,
                JLibDataGenerationTp.Instance,
                TypePackage.GetNested<MinimumCodeValueTypeIds>())
            .AddSingleton<ShoppingServiceMock>()
            .AddScopedAlias<IShoppingService, ShoppingServiceMock>()
            .AddAutoMapper(b => b.AddProfiles(typeCache, loggerFactory))
            .AddDataPackages(typeCache);

        var serviceProvider = serviceCollection
            .BuildServiceProvider()
            .DisposeWith(_disposables)
            .IncludeDataPackages<CustomerDp>();

        _shoppingService = serviceProvider.GetRequiredService<ShoppingServiceMock>();
    }
    public void Dispose()
        => _disposables.DisposeAll();


    /*************************************************************\
    |                            Test                             |
    \*************************************************************/
    [Fact]
    public void Test()
        => _shoppingService.Customers.MatchSnapshot();
}
