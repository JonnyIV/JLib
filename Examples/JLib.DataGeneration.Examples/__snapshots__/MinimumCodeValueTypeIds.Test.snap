{
  "CustomerId { Value = 274fa88f-8946-41eb-b302-ade8a94624a5 }": {
    "Id": {
      "Value": "274fa88f-8946-41eb-b302-ade8a94624a5"
    },
    "UserName": "JLib.DataGeneration.Examples.MinimumCodeValueTypeIds.CustomerDp.CustomerId"
  }
}
